pandas/io/common.py

```diff
@@ -322,18 +309,19 @@ def get_filepath_or_buffer(
             "storage_options passed with file object or non-fsspec file path"
         )
 
-    if isinstance(filepath_or_buffer, (str, bytes, mmap.mmap)):
-        return IOargs(
-            filepath_or_buffer=_expand_user(filepath_or_buffer),
-            encoding=encoding,
-            compression=compression,
-            should_close=False,
-            mode=mode,
-        )
+    if is_file_like(filepath_or_buffer) or isinstance(filepath_or_buffer, mmap.mmap):
+        return filepath_or_buffer, None, compression, False
 
-    if not is_file_like(filepath_or_buffer):
+    try:
+        return (
+            _expand_user(filepath_or_buffer, fallback=False),
+            None,
+            compression,
+            False,
+        )
+    except TypeError as err:
```

Comment:
instead of this try/except (and the fallback arg), can we just raise this inside _expand_user (and have this buble up)?